src/mlpack/methods/reinforcement_learning/eql_impl.hpp

```diff
@@ -0,0 +1,310 @@
+  ReplayType
+>::EQL(TrainingConfig& config,
+       NetworkType& network,
+       PolicyType& policy,
+       ReplayType& replayMethod,
+       UpdaterType updater,
+       EnvironmentType environment):
+    config(config),
+    learningNetwork(network),
+    policy(policy),
+    replayMethod(replayMethod),
+    updater(std::move(updater)),
+    #if ENS_VERSION_MAJOR >= 2
+    updatePolicy(NULL),
+    #endif
+    environment(std::move(environment)),
+    totalSteps(0),
+    deterministic(false)
+{
+  // To copy over the network structure.
+  targetNetwork = learningNetwork;
+
+  // Set up q-learning network.
+  if (learningNetwork.Parameters().is_empty())
+    learningNetwork.ResetParameters();
+
+  targetNetwork.ResetParameters();
+
+  #if ENS_VERSION_MAJOR == 1
+  this->updater.Initialize(learningNetwork.Parameters().n_rows,
+                           learningNetwork.Parameters().n_cols);
+  #else
+  this->updatePolicy = new typename UpdaterType::template
+      Policy<arma::mat, arma::mat>(this->updater,
+                                   learningNetwork.Parameters().n_rows,
+                                   learningNetwork.Parameters().n_cols);
+  #endif
+
+  // Initialize the target network with the parameters of learning network.
+  targetNetwork.Parameters() = learningNetwork.Parameters();
+}
+
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+EQL<
+  EnvironmentType,
+  NetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::~EQL()
+{
+  #if ENS_VERSION_MAJOR >= 2
+  delete updatePolicy;
+  #endif
+}
+
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+arma::uvec EQL<
+  EnvironmentType,
+  NetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::BestAction(const arma::mat& actionValues, const arma::mat& weightSpace)
+{
+  size_t numWeights = weightSpace.n_cols;
+  size_t actionSize = EnvironmentType::Action::size;
+  size_t batchSize = actionValues.n_cols / (numWeights * actionSize);
+  size_t extendedSize = numWeights * batchSize;
+
+  // Each preference vector is repeated batchSize * actionSize
+  // number of times. Shape: (rewardSize, extendedSize * actionSize).
+  const arma::mat extWeights = [&]()
+  {
+    arma::mat retval(rewardSize, extendedSize * actionSize);
+    size_t colIdx {}, start {};
+    size_t gap = batchSize * actionSize;
+
+    while (colIdx < numWeights)
+    {
+      retval.submat(arma::span(0, rewardSize),
+                    arma::span(start, start + gap - 1)) =
+          arma::repmat(weightSpace.col(colIdx), 1, gap);
+      start += gap;
+      ++colIdx;
+    }
+
+    return retval;
+  }();
+
+  // Batch multiply extWeights and actionValues and take the max index.
+  return arma::index_max(arma::reshape(arma::sum(extWeights % actionValues, 0),
+                                       actionSize, extendedSize), 0);
```

Comment:
```suggestion
  return arma::index_max(arma::reshape(arma::sum(extWeights % actionValues, 0),
      actionSize, extendedSize), 0);
```

Should be tabbed twice.

Reply:
Did this so that it's clear that those args are for ```arma::reshape```, the meaning would be lost if use the normal tabbings.